### ISSUE TITLE
Unable to update settings in ngHandsontable

### ISSUE THREAD

**Issue by garypma** [handsontable/ngHandsontable#156]
I have an AngularJs application that needs to show dynamic settings depending upon the user logged in. If I try to update the settings after a promise is returned with the needed data, the columns don't switch in the table as needed, they stay the default of A, B, C etc. Here is my code:

```
  var onGetUserComplete = function(data) {
        //$scope.settings = tmViewSettings(); // returns the hash object of settings required.
        if (userService.hasTmView()) {
            $scope.settings = tmViewSettings();            
            var table = hotRegisterer.getInstance('compTable');
            table.updateSettings(settings, true);

        } else if (userService.hasFgView()) {
            $scope.settings = fgHeadViewSettings();
        }

        loadGrid();
    };
```

Note the settings look something like this:

   var tmViewSettings = function () {
            var settings = {
                contextMenu: ['undo'],
                afterChange: $scope.itemChanged,
                hiddenColumns: {
                    columns: [0],
                    indicators: false
                },
                nestedHeaders: [
                  ['id', { colspan: 4 }, { label: 'Planning', colspan: 4 }, { colspan: 2 }, { label: 'Credits', colspan: 6 },
                      { colspan: 4 }, { label: 'Salary', colspan: 3 }, { colspan: 2 }, { label: 'Flags', colspan: 4 }],
                  ['id', 'Name', 'Practice', 'Country', 'Currency', 'YE15 PO', 'PM PO',
                      'Planning PO', 'Comp PO', 'PO for Calc', 'PO Communicated', 'Credits',
                      'D Credits', 'Pr Credits', 'Reason', 'Total Credits', 'Guaranteed TC', 'Total Comp',
                      'Bonus', '% Credit Inc', '% TC Inc', 'Current Salary', 'Grid Salary', 'D Salary',
                      'New Salary', '% Salary Inc', 'Promote', 'CC', 'Exp Hire', 'Transfer']
                ],
                collapsibleColumns: [
                  { row: -2, col: 5, collapsible: true },
                  { row: -2, col: 11, collapsible: true },
                  { row: -2, col: 21, collapsible: true },
                  { row: -2, col: 26, collapsible: true }
                ],
                filters: true,
                /_dropdownMenu : true,_/
                dropdownMenu: ['alignment', '----------', 'filter_by_condition', 'filter_action_bar'],
                columns: [
                  { data: 'rowId' },
                  { data: 'editLink', readOnly: true, renderer: "html", width: 130 },
                  { data: 'practice', readOnly: true, width: 80 },
                  { data: 'countryCode', readOnly: true, width: 80 },
                  { data: 'currencyCode', readOnly: true, width: 80 },
                  { data: 'lastYearPo', readOnly: true, width: 90 },
                  { data: 'pmpo', readOnly: true, width: 70 },
                  { data: 'tmpo', type: 'dropdown', source: loadPoOptions, width: 70 },
                  { data: 'compPO', type: 'dropdown', source: loadPoOptions, width: 80 },
                  { data: 'poForCalcs', readOnly: true, width: 90 },
                  { data: 'communicatedPo', type: 'dropdown', source: loadPoOptions, width: 120 },
                  { data: 'credits', type: 'numeric', readOnly: true, width: 70 },
                  { data: 'discretionaryCredits', type: 'numeric', width: 70 },
                  { data: 'fgPrincipalCredits', type: 'numeric', width: 80 },
                  { data: 'discretionaryReason', width: 100 },
                  { data: 'totalCredits', type: 'numeric', readOnly: true, width: 90 },
                  { data: 'guaranteedTc', type: 'numeric', format: '$ 0,0.00', width: 100 },
                  { data: 'totalComp', type: 'numeric', format: '$ 0,0.00', readOnly: true, width: 90 },
                  { data: 'bonus', type: 'numeric', format: '$ 0,0.00', readOnly: true, width: 90 },
                  { data: 'percentCreditIncrease', type: 'numeric', format: '0,0.00', readOnly: true, width: 100 },
                  { data: 'percentTcIncrease', type: 'numeric', format: '0,0.00', readOnly: true, width: 90 },
                  { data: 'currentSalary', type: 'numeric', format: '$ 0,0.00', readOnly: true, width: 100 },
                  { data: 'gridSalary', type: 'numeric', format: '$ 0,0.00', readOnly: true, width: 90 },
                  { data: 'discretionSalary', type: 'numeric', format: '$ 0,0.00', width: 90 },
                  { data: 'newSalary', type: 'numeric', format: '$ 0,0.00', readOnly: true, width: 90 },
                  { data: 'percentSalaryIncrease', type: 'numeric', format: '0,0.00', readOnly: true, width: 100 },
                  { data: 'promoteToPartner', type: 'checkbox', readOnly: true, width: 70 },
                  { data: 'careerConcern', type: 'checkbox', readOnly: true, width: 40 },
                  { data: 'experiencedHire', type: 'checkbox', readOnly: true, width: 70 },
                  { data: 'countryTransfer', type: 'checkbox', readOnly: true, width: 70 }
                ]
            };
            return settings;
        };

If I immediately set these setting to the table at the startup of the controller, the table works fine. If I wait for the callback to occur, then the settings are never applied, in fact they are ignored outright. If there is some solution for this it would be appreciated. Right now with the number of issues I have found with the angular version of Handsontable, I'm about ready to give up on it. It works great as long as you stay with the cookie cutter approach, but even using dropdowns don't work if you try to define them using <hot-column>. Please let me know on this as the whole angular version is getting really frustrating to work around.

**Comment by jprince** [comment]
Did you end up finding a solution to this? I'm running into the same issue.

**Comment by garypma** [comment]
No, they apparently don't respond to the angular plugin like they do the basic handsontable. I did however find a workaround:

First, set the settings property to just very basic items like callbacks. When the controller initializes, that way the grid doesn't try to hold on to any particular setting. Define a scope variable for each of the items you want to lazy load like columns, nested-headers, hidden-columns. My table looks like this:

<hot-table hot-id="compTable"
               datarows="participants"
               settings="{afterChange: itemChanged}"
               row-headers="false"
               col-headers="true"
               height="550"
               column-sorting
               dropdown-menu="dropdownMenu"
               columns="columns"
               nested-headers="nestedHeaders"
               collapsible-columns="collapsibleColumns"
               hidden-columns="hiddenColumns"
               filters>
    </hot-table>

Next, wait for the promise to return and then set the settings based upon the data returned in the promise, but not the wholes settings, each individual item from an object you have already defined like below:

           $scope.datalist = [];
            var settings = null;
            if (userService.hasTmView()) {
                settings = tmViewSettings();               
            } else if (userService.hasFgView()) {
                settings = fgHeadViewSettings();
            } else if (userService.hasAdminView()) {
                settings = null;
            }
            else {
                settings = groupViewerSettings();
            }
            if (settings !== null) {
                // Set all the individual setting properties as needed from the appropriate object.
                $scope.columns = settings.columns;
                $scope.nestedHeaders = settings.nestedHeaders;
                $scope.collapsibleColumns = settings.collapsibleColumns;
                $scope.hiddenColumns = settings.hiddenColumns;
                // now load the table with data
                loadGrid();
            }


In my case I have 4 or so predefined "settings" and based upon what is returned in my api call I get the correct settings. It's kind of hard coded for now, but you can get the idea here. What I don't do is set the settings property in the table, that still doesn't work and they need to fix it.

**Comment by jprince** [comment]
Bummer - thanks for the response.

**Comment by mudassir0909** [comment]
I think a better solution is to add `ng-if` on your `<hot-table>` directive

**Comment by garypma** [comment]
@mudassir0909 Not sure if that fixes my problem. Are you saying you used an ng-if to create a different hot table based upon some criteria, or are you saying you used an ng-if on the columns property on the hot table which I'm not sure you can do. 

If you have an example that explains what your doing I guess that might help, but I don't thing that works for my situation.

**Comment by mudassir0909** [comment]
### using ng-if
Assuming you are doing the following during the success callback of a async request(promise)

```js
promise.then(function(){
  //do calculations 
  $scope.settings = {
     //your dynamic settings
  };
});
```

Now adding `ng-if` in the html will ensure that the directive is compiled when the `$scope.settings` is available which means after your async request is done processing
```html
<hot-table ng-if="settings">
```

Makes sense?

### using $
using `$compile` gets little complicated, `ng-if` should solve your issue most of the times & I don't think `$compile` will have performance issues since it's what angular uses internally as well. The way `$compile` works is you can compile any html which has angular directives in the context of a given `$scope`

Assuming you have a directive which wraps `hot-table` under the link function you could do
```js
link: function(scope, element, attrs) {
  // A better option would be to use $templateCache to pull the template from the html file
  var html = "<hot-table> <!-- hot columns --> </hot-table>
  element.html($compile(html)(scope);
}
```
Basically you can trigger your wrapper directive based on conditions


Makes sense?

**Comment by garypma** [comment]
@mudassir0909 
I tried what you suggested with the ng-if, however it doesn't work. In fact the whole table never renders at all. I already have my table in a directive, so I may try your second option. What I have already works and is essentially what your doing except not using the ng-if. I have used ng-if before on other normal html components, but it doesn't seem to work on the hot table. I may be because my table is already in a directive, but I'm not sure.
